lib/Geography/MapAlgebra.hs

```diff
@@ -300,47 +316,95 @@ instance (Num a, KnownNat r, KnownNat c) => Num (Raster p r c a) where
+strict u (Raster a) = Raster $ computeAs u a
+{-# INLINE strict #-}
+
+-- | Create a `Raster` of any size which has the same value everywhere.
+constant :: (KnownNat r, KnownNat c, Construct u Ix2 a) => u -> Comp -> a -> Raster u p r c a
+constant u c a = fromFunction u c (const a)
+{-# INLINE constant #-}
+
+-- | \(\mathcal{O}(1)\). Create a `Raster` from a function of its row and column number respectively.
+fromFunction :: forall u p r c a. (KnownNat r, KnownNat c, Construct u Ix2 a) =>
+  u -> Comp -> (Ix2 -> a) -> Raster u p r c a
+fromFunction u c f = Raster $ makeArrayR u c sh f
+  where sh = fromInteger (natVal (Proxy :: Proxy r)) :. fromInteger (natVal (Proxy :: Proxy c))
+{-# INLINE fromFunction #-}
+
+-- | \(\mathcal{O}(1)\). Create a `Raster` from the values of any `GV.Vector` type.
 -- Will fail if the size of the Vector does not match the declared size of the `Raster`.
-fromUnboxed :: forall p r c a. (KnownNat r, KnownNat c, U.Unbox a) => U.Vector a -> Maybe (Raster p r c a)
-fromUnboxed v | (r * c) == U.length v = Just . Raster . R.delay $ R.fromUnboxed (R.ix2 r c) v
-              | otherwise = Nothing
+fromVector :: forall v p r c a. (KnownNat r, KnownNat c, GV.Vector v a, Mutable (A.ARepr v) Ix2 a, Typeable v) =>
+  Comp -> v a -> Maybe (Raster (A.ARepr v) p r c a)
+fromVector comp v | (r * c) == GV.length v = Just . Raster $ A.fromVector comp (r :. c) v
+                  | otherwise = Nothing
   where r = fromInteger $ natVal (Proxy :: Proxy r)
         c = fromInteger $ natVal (Proxy :: Proxy c)
 
--- | O(n). Create a `Raster` from a list of anything. Will fail if the size of the list
--- does not match the declared size of the `Raster`. In general, should be used for
--- debugging only.
-fromList :: forall p r c a. (KnownNat r, KnownNat c) => [a] -> Maybe (Raster p r c a)
-fromList l | (r * c) == length l = Just . Raster . R.delay $ R.fromListVector (R.ix2 r c) l
-           | otherwise = Nothing
-  where r = fromInteger $ natVal (Proxy :: Proxy r)
-        c = fromInteger $ natVal (Proxy :: Proxy c)
+-- | An RGBA image whose colour bands are distinct. Since each band starts as `D`,
+-- any band you don't use won't consume extra memory.
+data RGBARaster p r c a = RGBARaster { _red   :: Raster D p r c a
+                                     , _green :: Raster D p r c a
+                                     , _blue  :: Raster D p r c a
+                                     , _alpha :: Raster D p r c a }
+
+-- | Read any image type into a `Raster` of distinct colour bands
+-- with the cell type you declare. If the source image stores its
+-- values as `Int` but you declare `Double`, the conversion will happen
+-- automatically.
+--
+-- Will fail with `Left` if the declared size of the `Raster`
+-- does not match the actual size of the input image.
+fromRGBA :: forall p r c a. (Elevator a, KnownNat r, KnownNat c) => FilePath -> IO (Either String (RGBARaster p r c a))
```

Comment:
@lehins It's this function in particular that I'm talking about, re: automatic `Seq` vs `Par` detection. We're already in `IO`, so we could do `getNumCapabilities` to detect CPU core count. Otherwise, regardless of core count, if the image is sufficiently small we could just set it to `Seq`.

Reply:
`getNumCapabilities` isn't very relevant here, since using `Par` will use all capabilities under the hood even if we aren't in the `IO`. My honest guess though, that most of the time your library would be used on large rasters (bigger than 100x100 pixels), right? If that's the case, just don't even worry about `Seq` and always use `Par`, this way users will get the fastest performance, and possibly occasionally will take a tiny hit on small imagery. 
Otherwise you can try and experiment with benchmarks, see what's the golden number (image size) for you and go from there.

Reply:
Realistically yeah, the rasters would be large. Small would be 256 x 256, largest would be 60k x 60k.